lib/ansible/modules/clustering/kubevirt/kubevirt_vm.py

```diff
@@ -376,10 +376,7 @@ def execute_module(self):
                 ret = self.manage_state(state)
                 changed = changed or ret
 
-        self.exit_json(**{
-            'changed': changed,
-            'vm': result,
-        })
+        self.exit_json(**result)
```

Comment:
The reason wer return it like this is because we return proper changed also if running state was changed..

Reply:
Wait, so if you change `spec.running` from _true_ to _false_ or vice versa, then the `result` from `perform_action()` will not have `changed: true`? Why?

Reply:
so maybe instead of removing it we should fix the bug you mentioned above.

Reply:
Ah, now I've read the code. If you're running `patch_resource()` inside of `manage_state()`, then I think we should just override the `changed` value from `result` instead of embedding `result` one level down. It's not like that value is somehow incorrect.

Reply:
I've just pushed an (untested) version of what I've had in mind as the better solution to this.

Reply:
Yup, tested, works as I've intended. I believe this is a better approach of reporting results from this module. And in line with what upstream k8s and our other modules do.

Reply:
So please also update RETURN documentation appropriatelly

Reply:
Ok, copied the k8s_service RETURN docs, since that's exactly what we'll be returning as well.

Reply:
Yeah and now comes questions if we want to, I've created 'vm' return type, to hide implementation detail of VMI/VM, now user gets always different things. `items` are not supported here, I've always use first `resource_definition`. I am fine with this return type, but this *must* be backward comaptible so we must decide once and forever.

Reply:
Yup, you're right on your reasoning. I think I can get this to a state where we'll both be happy. I'll post a patch soonish.

Reply:
Well, this is simplification module, we must remember that. Most of the ansible modules return whatever it manages. For example see here:
  https://docs.ansible.com/ansible/latest/modules/stat_module.html#return-values

I think we really should just return 'vm' with all of it API attributes. So I believe we agree on this.

For the `k8s_objects`, not sure I follow. I think as part of vm dict we should return all vm attributes. But just a single VM. I think we shouldn't support multiple VM management in this module, but I am open changing this, if there is good reason, and not too much maintanence effort.

So IMHO, we should just return vm dict and inside it the vm attributes.

Reply:
even more likely, maybe we should define the return value `kubevirt_vm` not just `vm`.

Reply:
The point of the module is to abstract away the lowlevel stuff and make it seem like this is just a vm. Currently this means we're operating on both VMs and VMIs, yet returning only the former's results with some modifications (`changed`). In the future we might end up handling even more resource types and/or different versions of VMs/VMIs, which will return yet more different things. I don't think we should just assume that VM is the thing we'll be returning and what the user should expect to be getting.

I think that if we really want to be safe long term, we should ensure that the abstraction is properly handled on the RETURN values. So my idea is that we put the abstracted results directly under (following your suggestion) `kubevirt_vm` complex return structure and put any raw k8s objects we've touched during operation inside the `kubevirt_vm.k8s_objects` list. So the end form of RETURN might be something like this:

- `kubevirt_vm.primaryIP` – empty if stopped, taken from VMI if running.
- `kubevirt_vm.publicIP` – if we ever support ClusterIP Service, then that'd be the clusterIP; empty if not configured
- `kubevirt_vm.apiVersion` – if there are multiple VM versions, then that'd probably be it
- (…) – who knows what we'll get requests for as being useful to end users
- `kubevirt_vm.k8s_objects` – a list of low–level objects that we've used for those users that actually know their kubevirt and need to access a very specific VMI parameter, for example.

Granted, I'm not fluent enough in ansible to know how return values are used in practice. Maybe `k8s_objects` should be a dict and not a list, so that one can do `kubevirt_vm.k8s_objects.VMI.result.xyz`, but that implies we always operate on a single resource type – makes sense for now, but I'm not 100% certain it'll hold true for ever. The point is figuring out a way to return *everything* that might be potentially useful to end users.

Anyway, that's my reasoning for this change in the context of getting long term stable and flexible RETURN interface.

Reply:
tl;dr – I don't like the discrepancy between abstracting away VMs/VMIs on the input, but then neither actually abstracting it fully nor returning a full complement of low–level objects on the output. Instead I'd like us to do both.

Reply:
Ok now I understand what you've meant by `k8s_objects`. It makes sense. Just note, those most probably will be `kubevirt_objects`, but that's just cosmetic note.

Well, we can return it, no problem with that. I am just worried it would be too complicated.

Reply:
Putting my admin hat on: best software is designed in such a way, that doing basic stuff is trivial, yet arbitrarily complicated stuff is still achievable. Worst times as an admin are where the designer thought „there's *surely* no way anyone would need to do *that*”. Hence my approach to achieving both: curated high level return values should we need them and an easy access to all the raw low level results for the 1%.

That said, I've talked myself into switching from a list to a dict. If we ever need to return, say, multiple VMIs in `k8s_objects`, then we can always have lists like `k8s_objects.VMIlist` or `k8s_objects.VMIs`. I'll push a modified version in a bit.